changes/2020-06-04_how-to-fail-with-keyrings/change.md

```diff
@@ -0,0 +1,216 @@
+
+- Defining the decryption contract for each keyring
+  is out of scope.
+  We will address this in future specification changes.
+
+- A mechanism to interrogate a keyring
+  to determine the properties of its decryption contract
+  is out of scope.
+  We SHOULD address this in future changes.
+
+- This change only applies to keyrings.
+  We SHOULD define analogous concepts for
+  cryptographic materials managers
+  and the AWS Encryption SDK client
+  in future changes.
+
+## Motivation
+
+We lack a consistent definition both of
+what it means for a keyring to fail
+and what causes a keyring to fail.
+Without this consistent definition,
+different keyrings and implementations
+do slightly different things,
+leading to implementation drift
+and confusion for users and implementors.
+
+In order to clarify this situation,
+we MUST first define what it means for a keyring to fail,
+and that means defining what it means for a keyring to succeed.
+To do this,
+we define a new concept of the
+"keyring decryption contract".
+
+## Drawbacks
+
+This change SHOULD not introduce any drawbacks.
+It is simplifying and clarifying
+a relationship that already exists
+in the specification
+and clearly defining
+how that relationship affects behavior.
+
+## Security Implications
+
+The keyring decryption contract
+simplifies the definition of
+the security properties of a keyring
+and provides a consistent way to describe
+the security properties of each keyring.
+
+## Operational Implications
+
+Clearly defining what causes keyrings to fail
+makes it simpler for users to predict failing behavior
+based on either environment
+and across implementations.
+
+## Guide-level Explanation
+
+When a keyring succeeds on encryption,
+it creates an artifact: one or more encrypted data keys.
+This artifact has certain requirements that
+MUST be met on decryption
+in order to obtain the plaintext data key.
+This set of requirements is what we call
+the "keyring decryption contract."
+The process that a keyring uses to create this artifact
+is largely irrelevant to the decryption contract.
+What matters is the requirements
+that a decryption contract defines.
+
+## Reference-level Explanation
+
+### Definitions and Terms
+
+- **writing the decryption contract** :
+  This refers to the process of taking the plaintext data key
+  and successfully creating an artifact
+  (one or more encrypted data keys)
+  that requires a decrypting keyring to
+  fulfill the encrypting keyring's decryption contract.
+
+- **fulfilling the decryption contract** :
+  This refers to the process of taking one or more encrypted data keys
+  and successfully using them to obtain the plaintext data key.
+
+### How to Fail
+
+On encryption,
+if a keyring is unable or unwilling to
+write its decryption contract,
+that keyring MUST fail.
+
+On decryption,
+if a keyring is unable or unwilling to
+fulfill its decryption contract
+using the encrypted data keys that it receives,
```

Comment:
Does the decryption keyring have a contract?  Or is it trying/able to fulfill the contracts it receives? e.g. the encrypted data keys?

Reply:
> Does the decryption keyring have a contract?

No, but it does need to be able to determine if it SHOULD be able to fulfill a given decryption contract (EDK).

Reply:
Then what about this?

```suggestion
fulfill the decryption contract
of the encrypted data keys that it receives,
```

Reply:
+1, I think we should be consistent that "the decryption contract of a keyring" always refers to the decryption contract of the artifacts it produces when encrypting.